### ISSUE TITLE
constr:() is very slow in cases where it seems like it shouldn't be

### ISSUE THREAD

**Issue by JasonGross** [coq/coq#6545]
#### Version

8.8+alpha (e2d1c676b23a335b4fb8a528c99dfca2b82a1a39)

#### Operating system

Linux

#### Description of the problem

(See Also #3280)

What is `constr` spending its time doing?

```coq
Fixpoint twos (n : nat) (A : Set) :=
  match n with
  | 0 => A
  | S n' => twos n' (prod A A)
  end.

Tactic Notation "display" string(tacname) tactic3(tac) :=
  match goal with |- let n := ?k in _ => idtac k "Time" tacname end;
  lazy [twos];
  time tac ().

Notation "'hide'" := (_ = _) : type_scope.

Goal let n := 21 in twos n nat = twos n nat.
  display ""
          (fun _ => lazymatch goal with
                    | [ |- ?G ]
                      => let dummy := constr:(G) in
                         idtac
                    end). (* 0.319 s *) Undo.
  display ""
          (fun _ => lazymatch goal with
                    | [ |- ?f ?x ]
                      => let dummy := constr:(f x) in
                         idtac
                    end). (* 2.186 s *) Undo.
```

**Comment by ppedrot** [comment]
It's retyping the term.

**Comment by ppedrot** [comment]
More precisely, this seems to be due to a econstr-to-constr conversion in the type-checking of template-polymorphic inductive types. You're manipulating huge products, which are template-polymorphic, so you're hitting right in the algorithmic worst case.

**Comment by SkySkimmer** [comment]
I don't think it's about econstr, this seems solely about template polymorphism and retyping.
For instance if you do
~~~coq
Lemma foo: let n := 21 in twos n nat = twos n nat.
  lazy [twos].
  exact eq_refl.
Time  Qed.
~~~
it takes a while even if you avoid template poly by adding `Axiom prod : Set -> Set -> Set.` at the beginning.
However avoiding template poly makes the second `display` much quicker.
Basically retyping on a stack of template poly types is close to full typing.

**Comment by ppedrot** [comment]
> Basically retyping on a stack of template poly types is close to full typing.

Which is were true polymorphic inductive types shine... But you have to pay the cost at node creation instead.

**Comment by SkySkimmer** [comment]
Probably wontfix until the death of templates (may be correlated with the death of the universe)
